tests/test_apply.py

```diff
@@ -75,7 +116,19 @@ def deploy_pulumi_script(script: str, version: str, select_stack: bool, select_c
         if select_cwd and not should_fail:
             cmd.extend(["--cwd", temp_dir])
         out = run(cmd, **kwargs)
-        return out[1]
+        return temp_dir
```

Comment:
I think this function should be made into a context manager, as when the temp dir is returned, `tempfile.TemporaryDirectory` will be cleaned up. This may run the cleanup at garbage collection time, which might be why this seems to work.

Changes:
* add 'from contextlib import contextmanager' to the imports
* add `@contextmanager' decorator to the `deploy_pulumi_script` function
* replace this `return` with `yield`

This might mean you have to do the same with `create_dummy_stack`, and use a `with` block at the call sites. Something to think about

Reply:
You are right in general, however this is a really specific usage here. I use this return value only once in the tests and that is a valid case with a correct behaviour since `tempfile.TemporaryDirectory` with Python 3.12 adds a `delete` argument. If that's been set to `False`, as a result the temp directory is not deleted and can be processed and checked. Not the nicest, fail-safest solution though I agree, but does the job.

Reply:
The way you have it currently you are setting `cleanup=False` in `create_dummy_stack` which as you say is not cleaned up, so the test will execute fine, however that leaves the temporary directory around. As you say this leaves the directory around after the test if someone wants to inspect it, but I feel the better behaviour is to cleanup the dir after the test, and the developer can change the `cleanup` flag manually if they are making changes. Or add a testing configuration envar to control the behaviour.

This is a fairly minor point so I'll leave the final call up to you.

Reply:
As an interim solution I add a `try...finally` to the test in question, all other cases are handled with `with`.